engine/common/synchronization/engine.go

```diff
@@ -509,8 +509,72 @@ func (e *Engine) sendRequests(participants flow.IdentifierList, ranges []chainsy
 	}
 }
 
-// TODO: implement spam reporting similar to validateSyncRequestForALSP
-func (e *Engine) validateBatchRequestForALSP(channel channels.Channel, id flow.Identifier, batchRequest *messages.BatchRequest) (*alsp.MisbehaviorReport, bool, error) {
+// validateBatchRequestForALSP checks if a batch request should be reported as a misbehavior due to malicious intent (e.g. spamming).
+// It returns a misbehavior report and a boolean indicating whether validation passed, as well as an error.
+// Returns an error that is assumed to be irrecoverable because of internal processes that didn't allow validation to complete.
+// Returns true if the batch request is valid and should not be reported as misbehavior.
+// Returns false if either
+// a) the batch request is invalid or
+// b) the batch request is valid but should be reported as misbehavior anyway (due to probabilities) or
+// c) an error is encountered.
+func (e *Engine) validateBatchRequestForALSP(originID flow.Identifier, batchRequest *messages.BatchRequest) (*alsp.MisbehaviorReport, bool, error) {
```

Comment:
In this function (also for `RangeRequest`), the "load factor" (probability of reporting an ambiguous misbehaviour) is proportional to the number of blocks requested. However, the request handler will cap the number of blocks in the request that it actually attempts to retrieve at the configured max request limit (64 by default). 

- For range requests, the message size is constant. Therefore the actual load associated with a request for blocks 100-164 and for blocks 100-10000 is the same.
- For batch requests, the message size is proportional to the number of requested blocks. So larger requests introduces some very small additional load, on memory use and the serialization, that is linear w.r.t. the number of requested blocks. But it's tiny compared to the cost of reading the blocks we _do_ service from the database. 

Nodes should not be sending requests in excess of the limit anyway, so it doesn't really matter. But it would be more correct to cap the the "load factor" once we reach the maximum request size.

Reply:
It's best to ignore `MaxSize` in the early pipeline stages. Truncating big requests [happens later at the `RequestHandler` stage]((https://github.com/onflow/flow-go/blob/ff19b11461ef50c0995d4ec2e634d8b24f5bea72/engine/common/synchronization/request_handler.go#L290-L292)). Making decisions about ALSP reports earlier while considering `MaxSize` can create hidden dependencies between the sync engine and the `RequestHandler`. This goes against the principle of separation of concerns, making the code harder to manage, debug, and update.